# Zjazd 5, część 1 — Ćwiczenia laboratoryjne
## Temat: Pandas — Series, DataFrame, eksploracja danych

In [10]:
import pandas as pd
import numpy as np

---
## Ćwiczenie 1: Series — tworzenie i operacje

In [11]:
# Series z listy — przykład ze skryptu
oceny = pd.Series([4.5, 3.0, 5.0, 3.5, 4.0])
print(oceny)

0    4.5
1    3.0
2    5.0
3    3.5
4    4.0
dtype: float64


### Zadanie 1a — Series `pensje` ze słownika

In [12]:
pensje = pd.Series({
    'Anna':  5500,
    'Jan':   7200,
    'Ewa':   4800,
    'Marek': 9100,
    'Kasia': 6300
})
print(pensje)

# Sprawdzenie: pensje['Marek'] powinno być 9100
print("\npensje['Marek'] =", pensje['Marek'])

Anna     5500
Jan      7200
Ewa      4800
Marek    9100
Kasia    6300
dtype: int64

pensje['Marek'] = 9100


### Zadanie 1b — Odpowiedzi na pytania o `pensje`

In [13]:
# 1. Imię osoby zarabiającej najwięcej
print("Osoba zarabiająca najwięcej:", pensje.idxmax())

# 2. Średnia pensja
srednia = pensje.mean()
print("Średnia pensja:", srednia)

# 3. Osoby zarabiające powyżej średniej
powyzej_sredniej = pensje[pensje > srednia]
print("\nOsoby zarabiające powyżej średniej:")
print(powyzej_sredniej)

# 4. Liczba osób zarabiających poniżej 6000
ponizej_6000 = (pensje < 6000).sum()
print("\nLiczba osób zarabiających poniżej 6000 zł:", ponizej_6000)

Osoba zarabiająca najwięcej: Marek
Średnia pensja: 6580.0

Osoby zarabiające powyżej średniej:
Jan      7200
Marek    9100
dtype: int64

Liczba osób zarabiających poniżej 6000 zł: 2


### Zadanie 1c — Operacje na Series (podwyżki, premie)

In [14]:
# 1. Pensje po podwyżce 10%
pensje_podwyzka = pensje * 1.10
print("Pensje po podwyżce 10%:")
print(pensje_podwyzka)

# 2. Pensje po premii 500 zł
pensje_premia = pensje + 500
print("\nPensje po premii 500 zł:")
print(pensje_premia)

# 3. Nowa pensja = (pensja * 1.10) + 500
nowe_pensje = (pensje * 1.10) + 500
print("\nNowe pensje (podwyżka + premia):")
print(nowe_pensje)

# 4. Ranking od najwyższej
print("\nRanking nowych pensji (malejąco):")
print(nowe_pensje.sort_values(ascending=False))

# Sprawdzenie: nowa pensja Anny = 5500 * 1.10 + 500 = 6550.0
print("\nNowa pensja Anny:", nowe_pensje['Anna'])

Pensje po podwyżce 10%:
Anna      6050.0
Jan       7920.0
Ewa       5280.0
Marek    10010.0
Kasia     6930.0
dtype: float64

Pensje po premii 500 zł:
Anna     6000
Jan      7700
Ewa      5300
Marek    9600
Kasia    6800
dtype: int64

Nowe pensje (podwyżka + premia):
Anna      6550.0
Jan       8420.0
Ewa       5780.0
Marek    10510.0
Kasia     7430.0
dtype: float64

Ranking nowych pensji (malejąco):
Marek    10510.0
Jan       8420.0
Kasia     7430.0
Anna      6550.0
Ewa       5780.0
dtype: float64

Nowa pensja Anny: 6550.000000000001


---
## Ćwiczenie 2: DataFrame — tworzenie i atrybuty

### Krok 1 — DataFrame z dict

In [15]:
sklep = pd.DataFrame({
    'produkt':         ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Monitor'],
    'cena':            [3500, 1800, 2500, 350, 2200],
    'stan_magazynowy': [45, 120, 200, 500, 75],
    'kategoria':       ['Komputery', 'Mobilne', 'Mobilne', 'Akcesoria', 'Komputery']
})
sklep

,produkt,cena,stan_magazynowy,kategoria
0,Laptop,3500,45,Komputery
1,Tablet,1800,120,Mobilne
2,Smartfon,2500,200,Mobilne
3,Słuchawki,350,500,Akcesoria
4,Monitor,2200,75,Komputery


### Zadanie 2a — Eksploracja DataFrame `sklep`

In [16]:
# 1. Kształt tabeli
print("Shape:", sklep.shape)

# 2. Nazwy kolumn
print("Kolumny:", list(sklep.columns))

# 3. Typy danych
print("\nTypy danych:")
print(sklep.dtypes)

# 4. Pierwsze 3 wiersze
print("\nPierwsze 3 wiersze:")
sklep.head(3)

Shape: (5, 4)
Kolumny: ['produkt', 'cena', 'stan_magazynowy', 'kategoria']

Typy danych:
produkt              str
cena               int64
stan_magazynowy    int64
kategoria            str
dtype: object

Pierwsze 3 wiersze:


,produkt,cena,stan_magazynowy,kategoria
0,Laptop,3500,45,Komputery
1,Tablet,1800,120,Mobilne
2,Smartfon,2500,200,Mobilne


### Zadanie 2b — Nowe kolumny

In [17]:
# 1. Wartość magazynu
sklep['wartosc_magazynu'] = sklep['cena'] * sklep['stan_magazynowy']

# 2. Cena z VAT
sklep['cena_z_vat'] = sklep['cena'] * 1.23

# 3. Wyświetl cały DataFrame
print("DataFrame z nowymi kolumnami:")
print(sklep)

# Sprawdzenie: wartość magazynu Laptop = 3500 × 45 = 157500
print("\nWartość magazynu Laptop:", sklep.loc[sklep['produkt'] == 'Laptop', 'wartosc_magazynu'].values[0])

DataFrame z nowymi kolumnami:
     produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_vat
0     Laptop  3500               45  Komputery            157500      4305.0
1     Tablet  1800              120    Mobilne            216000      2214.0
2   Smartfon  2500              200    Mobilne            500000      3075.0
3  Słuchawki   350              500  Akcesoria            175000       430.5
4    Monitor  2200               75  Komputery            165000      2706.0

Wartość magazynu Laptop: 157500


### Krok 2 — Wczytanie CSV (`tips`)

In [18]:
url_tips = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv'
tips = pd.read_csv(url_tips)

### Zadanie 2c — Eksploracja `tips`

In [19]:
# 1. Shape
print("Shape:", tips.shape)

# 2. Ostatnie 5 wierszy
print("\nOstatnie 5 wierszy:")
print(tips.tail())

# 3. 3 losowe wiersze
print("\n3 losowe wiersze:")
print(tips.sample(3, random_state=42))

# 4. Typy danych
print("\nTypy danych:")
print(tips.dtypes)

Shape: (244, 7)

Ostatnie 5 wierszy:
     total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

3 losowe wiersze:
     total_bill   tip   sex smoker  day    time  size
24        19.82  3.18  Male     No  Sat  Dinner     2
6          8.77  2.00  Male     No  Sun  Dinner     2
153       24.55  2.00  Male     No  Sun  Dinner     4

Typy danych:
total_bill    float64
tip           float64
sex               str
smoker            str
day               str
time              str
size            int64
dtype: object


---
## Ćwiczenie 3: Eksploracja datasetu `penguins` — EDA

In [20]:
url_penguins = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv'
penguins = pd.read_csv(url_penguins)

### Zadanie 3a — Pierwszy rzut oka

In [21]:
# 1. Pierwsze 10 wierszy
print("head(10):")
penguins.head(10)

head(10):


,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,MALE
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,FEMALE
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,FEMALE
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,FEMALE
5,Adelie,Torgersen,39.3,20.6,190.0,3650.0,MALE
6,Adelie,Torgersen,38.9,17.8,181.0,3625.0,FEMALE
7,Adelie,Torgersen,39.2,19.6,195.0,4675.0,MALE
8,Adelie,Torgersen,34.1,18.1,193.0,3475.0,NaN
9,Adelie,Torgersen,42.0,20.2,190.0,4250.0,NaN


In [22]:
# 2. Shape
print("Shape:", penguins.shape)

Shape: (344, 7)


In [23]:
# 3. info() — brakujące wartości
penguins.info()

<class 'pandas.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    str    
 1   island             344 non-null    str    
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    str    
dtypes: float64(4), str(3)
memory usage: 18.9 KB


### Zadanie 3b — Statystyki opisowe

In [24]:
# 1. describe() — tylko kolumny numeryczne
print("describe():")
penguins.describe()

describe():


,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


In [25]:
# 2. describe(include='all') — wszystkie kolumny
print("describe(include='all'):")
penguins.describe(include='all')

describe(include='all'):


,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
count,344,344,342.000000,342.000000,342.000000,342.000000,333
unique,3,3,NaN,NaN,NaN,NaN,2
top,Adelie,Biscoe,NaN,NaN,NaN,NaN,MALE
freq,152,168,NaN,NaN,NaN,NaN,168
mean,NaN,NaN,43.921930,17.151170,200.915205,4201.754386,NaN
std,NaN,NaN,5.459584,1.974793,14.061714,801.954536,NaN
min,NaN,NaN,32.100000,13.100000,172.000000,2700.000000,NaN
25%,NaN,NaN,39.225000,15.600000,190.000000,3550.000000,NaN
50%,NaN,NaN,44.450000,17.300000,197.000000,4050.000000,NaN
75%,NaN,NaN,48.500000,18.700000,213.000000,4750.000000,NaN


In [26]:
# 3. Mediana masy ciała
mediana_masy = penguins['body_mass_g'].median()
print("Mediana masy ciała (body_mass_g):", mediana_masy, "g")

Mediana masy ciała (body_mass_g): 4050.0 g


### Zadanie 3c — Rozkłady wartości

In [27]:
# 1. Pingwiny wg gatunku
print("Liczba pingwinów wg gatunku:")
print(penguins['species'].value_counts())

# 2. Pingwiny wg wyspy
print("\nLiczba pingwinów wg wyspy:")
print(penguins['island'].value_counts())

# 3. Rozkład płci
print("\nRozkład płci:")
print(penguins['sex'].value_counts())

# 4. Procent gatunku Adelie
print("\nProcent każdego gatunku:")
print(penguins['species'].value_counts(normalize=True).round(3) * 100)

Liczba pingwinów wg gatunku:
species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

Liczba pingwinów wg wyspy:
island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

Rozkład płci:
sex
MALE      168
FEMALE    165
Name: count, dtype: int64

Procent każdego gatunku:
species
Adelie       44.2
Gentoo       36.0
Chinstrap    19.8
Name: proportion, dtype: float64


### Zadanie 3d — Brakujące dane

In [28]:
# 1. Brakujące wartości w każdej kolumnie
print("Brakujące wartości w każdej kolumnie:")
print(penguins.isna().sum())

# 2. Procent braków w kolumnie 'sex'
braki_sex = penguins['sex'].isna().sum()
procent_sex = braki_sex / len(penguins) * 100
print(f"\nBraki w kolumnie 'sex': {braki_sex} ({procent_sex:.1f}%)")

Brakujące wartości w każdej kolumnie:
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

Braki w kolumnie 'sex': 11 (3.2%)


### Zadanie 3e — Wnioski z EDA

1. **Dataset zawiera 344 pingwiny** z 3 gatunków (Adelie, Chinstrap, Gentoo) obserwowanych na 3 wyspach Antarktyki.
2. **Najczęstszy gatunek to Adelie** — 152 osobniki, co stanowi ok. 44,2% wszystkich pingwinów w datasecie.
3. **Brakuje danych w 5 kolumnach** — kolumny pomiarowe (dziób, płetwa, masa) mają po 2 braki każda, natomiast kolumna `sex` ma aż 11 brakujących wartości (3,2% danych).
4. **Średnia masa ciała wynosi ok. 4202 g, a mediana 4050 g** — lekka asymetria sugeruje, że niewielka grupa ciężkich osobników (gatunek Gentoo) podciąga średnią w górę.
5. **Pingwiny żyją na 3 wyspach; najliczniej zamieszkują Biscoe** (168 osobników), a najmniej liczna jest wyspa Torgersen (52 osobniki).
6. **Rozkład płci jest zbliżony do zrównoważonego** — 165 samic i 168 samców wśród osobników z określoną płcią, co stanowi niemal idealny podział 50/50.

---
## Ćwiczenie 4: Pytania biznesowe — dataset `tips`

In [29]:
# Dodaj kolumnę z procentem napiwku
tips['tip_pct'] = (tips['tip'] / tips['total_bill'] * 100).round(1)
tips.head()

,total_bill,tip,sex,smoker,day,time,size,tip_pct
0,16.99,1.01,Female,No,Sun,Dinner,2,5.9
1,10.34,1.66,Male,No,Sun,Dinner,3,16.1
2,21.01,3.50,Male,No,Sun,Dinner,3,16.7
3,23.68,3.31,Male,No,Sun,Dinner,2,14.0
4,24.59,3.61,Female,No,Sun,Dinner,4,14.7


### Zadanie 4a — Odpowiedzi na pytania biznesowe

In [30]:
# 1. Średni rachunek
print("Średni rachunek (total_bill):", round(tips['total_bill'].mean(), 2), "USD")

# 2. Średni napiwek w procentach
print("Średni napiwek (tip_pct):", round(tips['tip_pct'].mean(), 1), "%")

# 3. Dzień z największą liczbą rachunków
print("\nLiczba rachunków wg dnia:")
print(tips['day'].value_counts())
print("Dzień z najwięcej rachunkami:", tips['day'].value_counts().idxmax())

# 4. Rachunki od palaczy
print("\nLiczba rachunków wg palacza/niepalacza:")
print(tips['smoker'].value_counts())

# 5. describe() dla 'tip' — mediana
print("\nStatystyki kolumny 'tip':")
print(tips['tip'].describe())
print("Mediana napiwku:", tips['tip'].median(), "USD")

# 6. Procent rachunków z obiadu (Dinner)
print("\nRozkład pory posiłku (%):")  
print((tips['time'].value_counts(normalize=True) * 100).round(1))

# 7. Unikalne rozmiary grup
print("\nLiczba unikalnych rozmiarów grup:", tips['size'].nunique())
print("Dostępne rozmiary:", sorted(tips['size'].unique()))

Średni rachunek (total_bill): 19.79 USD
Średni napiwek (tip_pct): 16.1 %

Liczba rachunków wg dnia:
day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64
Dzień z najwięcej rachunkami: Sat

Liczba rachunków wg palacza/niepalacza:
smoker
No     151
Yes     93
Name: count, dtype: int64

Statystyki kolumny 'tip':
count    244.000000
mean       2.998279
std        1.383638
min        1.000000
25%        2.000000
50%        2.900000
75%        3.562500
max       10.000000
Name: tip, dtype: float64
Mediana napiwku: 2.9 USD

Rozkład pory posiłku (%):
time
Dinner    72.1
Lunch     27.9
Name: proportion, dtype: float64

Liczba unikalnych rozmiarów grup: 6
Dostępne rozmiary: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6)]
